utilities/ldap_sync.sql

```diff
@@ -25,118 +152,763 @@ Changes in this version:
+CREATE OR REPLACE PYTHON3 SCALAR SCRIPT "GET_AD_ATTRIBUTE" ("LDAP_CONNECTION" VARCHAR(2000),"SEARCH_STRING" VARCHAR(2000) UTF8, "ATTR"  VARCHAR(1000),  "VERIFY" VARCHAR(10))
+EMITS ("SEARCH_STRING" VARCHAR(2000) UTF8, "ATTR" VARCHAR(1000), "VAL" VARCHAR(1000) UTF8) AS
+#============================================================================================================================================================================
+'''
+###############################################
+--------------- General READ ME  --------------
+###############################################
+Note: This release is now PYTHON3
+
+Purpose: This script is designed to extract LDAP information.
+About:   This version of the script underwent a refactoring and was upgraded to Python3.       
+
+What is new in this version
+--------------------------------------------
+1. Added additional tests to validate the Connnection.
+---------------------------------------------
+  This script will end abnormally if a valid LDAP connection can not be made. Primarily,
+  this will aid troubleshooting when doing a first time implentation of the LDAP Sync
+  (think firewalls and forgetting to set the ldap parameter in EXAOperation).
+      This script is now more aggessive with validations and no longer passively returns
+  if valid LDAP data is not extracted, unless it is intentional.
+---------------------------------------------
+2. Conversion to Python3 includes:
+---------------------------------------------
+   a. Changes in handling string and byte datatypes. There are differences between 
+   Python2 and Python3. What used to work in Python2 no longer works as-is in Python3.
+   b. Accounting for new error message formats in Python3.
+---------------------------------------------
+3. Add additional documentation
+---------------------------------------------
+   To improve user experience with first time implementation
+   and troubleshooting.
+'''
+
 import ldap
+import socket           # To test LDAP connectivity and abnormally end if LDAP server is not reachable.
+                        # Do not return to calling script with null results (actually empty string)
+#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Script Variables
+#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+DEBUG = False               # Flag to scrutinize the CONNECTION
+results = ''                # Here is my empty string for storing LDAP data
+connect_result = ''         # Hold sock.connect_ex results
+abortMessage = ">>> Aborting with no action taken!"
+ldap_server =''             # Variable holding the input of ctx.LDAP_CONNECTION
+server = ''                 # Variable holding the extracted LDAP Server address
+port = 'null'               # Variable holding the extract LDAP Port
+uri=''                      # Varaible to hold input URI (url)
+colon = ":"                 #IPv4 / DNS variable containing a colon
+ipVersion4 = False          # Socket criteria using ipv4 address
```

Comment:
How about using the very same case for ipv4 / IPV4 / ... (and 6 as well) everywhere?